crates/turborepo-lib/src/run/summary/spaces.rs

```diff
@@ -1,19 +1,71 @@
+use std::{
+    fmt,
+    fmt::{Debug, Formatter},
+    time::Duration,
+};
+
 use chrono::{DateTime, Local};
 use serde::Serialize;
-use turbopath::AbsoluteSystemPathBuf;
-use turborepo_api_client::{APIAuth, APIClient};
+use tokio::{sync::mpsc::Sender, task::JoinHandle};
+use turborepo_api_client::{spaces::SpaceRun, APIAuth, APIClient};
+
+use crate::run::summary::{Error, RunSummary};
 
-struct SpacesClient {
+pub struct SpacesClient {
     space_id: String,
     api_client: APIClient,
     api_auth: APIAuth,
+    request_timeout: Duration,
+    errors: Vec<Error>,
 }
 
-#[derive(Serialize)]
-struct SpacesClientSummary {
-    id: String,
-    name: String,
-    version: String,
+/// Once the client is done, we return any errors
+/// and the SpaceRun struct
+pub struct SpacesClientResult {
+    pub errors: Vec<Error>,
+    // Can be None because SpacesClient could error on join
+    pub run: Option<SpaceRun>,
+}
+
+/// Handle on the space client, lets you send SpaceRequests to the worker
+/// thread and eventually await on the worker thread to finish
+pub struct SpacesClientHandle {
+    handle: JoinHandle<SpacesClientResult>,
+    tx: Sender<SpaceRequest>,
+}
+
+impl Debug for SpacesClientHandle {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        // We can't print much more than that since handle/tx are both
+        // opaque
+        f.debug_struct("SpaceClientHandle").finish()
+    }
+}
```

Comment:
In the past I've used `Derivate` for this sort of thing: https://mcarton.github.io/rust-derivative/latest/index.html